### DISCUSSION TITLE
HA Intents

### DISCUSSION THREAD

**Discussion by jaswalters** [toverainc/willow#168] [General]
If I create an intent and intent script with custom sentences for devices in HA, will Willow leverage these to alter a device?

**Comment by nikito** [comment]
Yes, I have numerous custom intents and intent scripts in HA that work with willow. Just note you need to be using WIS (not local commands) for it to work. Also in current code if your speech responses exceed 64 characters they will get cut off in tts, but this is being looked into in issue #159  😊

**Reply by jaswalters** [reply]
So just using an EPS32Box with HA will not work with intents as its not leveraging WIS correct?

**Reply by nikito** [reply]
Yes, so the ESP32Box has local commands, but they are presently limited to 400 total commands (200 on and 200 off), so mainly just device control. This is detailed here: https://github.com/toverainc/willow/wiki/Local-Command-Support

**Reply by kristiankielhofner** [reply]
The built-in HA intents work with either local commands or WIS. @nikito is doing some very cool things with his custom intents that absolutely require WIS because they require a lot of dynamic input/output. Examples:

"Hi ESP, turn on kitchen lights"

Local commands and WIS support this and will send the text to Home Assistant. Because lights are included in the intents in HA, if you have a light entity name kitchen it will match and execute.

"Hi ESP, what is the weather in [some random place]?"

Local commands have to be predefined ahead of time and there is no way it can match cities, entities, etc that have not been specifically configured. This will require WIS for speech to text.

Either local commands or WIS speech recognition mode support using WIS for text to speech output so in either case if you have a matching intent in HA, you are configured for TTS output, and there is some text output from HA Willow supports speaking these results back to you. There is also working being done by Espressif on local device text to speech but we don't have an ETA for that (they currently only support on device text to speech for Chinese).

**Reply by jaswalters** [reply]
TY! I tested and it worked with my intents with no issues.

**Reply by rob-odwyer** [reply]
How do you set up an intent like
> Hi ESP, what is the weather in [some random place]?

on the HA side?

No issues with detecting the arbitrary text with the demo WIS, but from what I can tell the HA sentence format seems to require that all slots in the sentence get filled with either a device name or a hardcoded replacement.
Is it possible to set it up to allow an arbitrary sequence of words after the initial phrase that matches, and handle those words in a script?

**Reply by nikito** [reply]
It's possible, but requires some custom intent sentence/script coding. They details this at https://developers.home-assistant.io/docs/voice/intent-recognition/template-sentence-syntax https://www.home-assistant.io/integrations/conversation/ and https://www.home-assistant.io/integrations/intent_script/

To do the above example, you would likely need to have some collection of locations stored in a list; I don't know that home assistant allows "wild card" words/slots that take in any random value.

**Reply by rob-odwyer** [reply]
OK cool - that lines up with what I'd seen so far.

I can define sentences and intent_scripts and get Willow to match them, but the "wild card" part is specifically what I was hoping to do.

The weather example isn't a great one since defining a list of locations isn't that bad - specifically what I want to do is add items to a grocery list.

Sounds like I'll either need to wait for support in the hassil language or build some kind of proxy in front of my HA that handles these commands instead.

**Reply by kristiankielhofner** [reply]
The "proxy in front of HA" is one of the use cases for the Willow Application Server (in development). The first release will only support management of Willow devices but it will move far beyond that.

There are a lot of voice assistant tasks that don't make any sense for Home Assistant to be involved in - such as your use case of adding items to a grocery list. Or things like calendars, reminders, timers, alarms, and many other (arguably most) tasks.

We hope with WAS and the functionality enabled by Willow and WIS we'll be able to kick-off a very robust open voice application and integration ecosystem.

The general concept is to use NLP and routing to applications and integrations - basically Alexa skills without the awkward and confusing "Ask XYZ" grammar where the user can define general grammar and inputs/outputs to various providers.

Full functionality for this is likely months away but it is very much on our radar.

**Reply by rob-odwyer** [reply]
That makes a lot of sense as a long term direction, thanks for the explanation!

Are you thinking of designing WAS to use a more specific grammar, e.g. certain sentence templates map to certain integrations?

Or something LLM-powered, like the way Langchain and similar tools construct agents? It would be interesting to use an agent that you could pass possible integrations to as actions along with a query. Might not be appropriate for everyone's use case though, since its inherently a bit unpredictable.

**Reply by kristiankielhofner** [reply]
WAS will use NLP approaches - LLMs don't really make sense for this use case and are impractical to run on most GPUs and CPUs.

We will likely use Rasa (or similar) for flexibility in grammar and defining of rules for intent matching to actions, apps, integrations, etc.
